core/services/ocrcommon/data_source.go

```diff
@@ -225,21 +233,45 @@ type inMemoryDataSourceCache struct {
 	// Even if updates fail, previous values are returned.
 	cacheFreshness  time.Duration
 	mu              sync.RWMutex
+	chStop          services.StopChan
+	chDone          chan struct{}
 	latestUpdateErr error
 	latestTrrs      pipeline.TaskRunResults
 	latestResult    pipeline.FinalResult
 	kvStore         job.KVStore
 }
 
+func (ds *inMemoryDataSourceCache) Start(context.Context) error {
+	go func() { ds.updater() }()
+	return nil
+}
+
+func (ds *inMemoryDataSourceCache) Close() error {
+	close(ds.chStop)
+	<-ds.chDone
+	return nil
+}
+
 // updater periodically updates data source cache.
 func (ds *inMemoryDataSourceCache) updater() {
 	ticker := time.NewTicker(ds.cacheFreshness)
-	for ; true; <-ticker.C {
-		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	ctx, cancel := ds.chStop.CtxCancel(context.WithTimeout(context.Background(), time.Second*10))
+	defer cancel()
+	updateCache := func() {
```

Comment:
We need it per-call thought, right?
```suggestion
	updateCache := func() {
		ctx, cancel := ds.chStop.CtxCancel(context.WithTimeout(context.Background(), time.Second*10))
		defer cancel()
```

Reply:
Changed, got confused for a second, forgot that context timeout is not reusable